port/raspberrypi/rp2xxx/src/hal/cyw43_pio_spi.zig

```diff
@@ -0,0 +1,227 @@
+//! Code based on embassy cyw43-pio driver https://github.com/embassy-rs/embassy/tree/main/cyw43-pio
+const std = @import("std");
+const microzig = @import("microzig");
+const hal = @import("../hal.zig");
+
+const Cyw43_Spi = microzig.drivers.wireless.Cyw43_Spi;
+
+const chip = microzig.hal.compatibility.chip;
+
+const cyw43spi_program = blk: {
+    @setEvalBranchQuota(5000);
+    break :blk hal.pio.assemble(
+        \\.program cyw43spi
+        \\.side_set 1
+        \\
+        \\.wrap_target
+        \\
+        \\; write out x-1 bits
+        \\lp:
+        \\out pins, 1    side 0
+        \\jmp x-- lp     side 1
+        \\
+        \\; switch directions
+        \\set pindirs, 0 side 0
+        \\nop            side 0
+        \\
+        \\; read in y-1 bits
+        \\lp2:
+        \\in pins, 1     side 1
+        \\jmp y-- lp2    side 0
+        \\
+        \\; wait for event and irq host
+        \\wait 1 pin 0   side 0
+        \\irq 0          side 0
+        \\
+        \\.wrap
+    , .{}).get_program_by_name("cyw43spi");
+};
+
+pub fn init(pio: hal.pio.Pio, cs_pin: hal.gpio.Pin, io_pin: hal.gpio.Pin, clk_pin: hal.gpio.Pin) !Cyw43PioSpi {
+    const sm = try pio.claim_unused_state_machine();
+
+    // Chip select pin setup
+    cs_pin.set_function(.sio);
+    cs_pin.set_direction(.out);
+    cs_pin.put(1);
+
+    // IO pin setup
+    io_pin.set_function(get_pin_pio_function(pio));
+    io_pin.set_output_disabled(false);
+    io_pin.set_pull(.disabled);
+    io_pin.set_schmitt_trigger(.enabled);
+
+    pio.set_input_sync_bypass(@truncate(@intFromEnum(io_pin)));
+
+    io_pin.set_drive_strength(.@"12mA");
+    io_pin.set_slew_rate(.fast);
+
+    // Clock pin setup
+    clk_pin.set_function(get_pin_pio_function(pio));
+    clk_pin.set_output_disabled(false);
+    clk_pin.set_drive_strength(.@"12mA");
+    clk_pin.set_slew_rate(.fast);
+
+    try pio.sm_load_and_start_program(sm, cyw43spi_program, .{
+        // Default max value from pico-sdk 62.5Mhz
+        .clkdiv = .{ .int = 2, .frac = 0 },
+        .pin_mappings = .{
+            .out = .{ .base = @truncate(@intFromEnum(io_pin)), .count = 1 },
+            .set = .{ .base = @truncate(@intFromEnum(io_pin)), .count = 1 },
+            .side_set = .{ .base = @truncate(@intFromEnum(clk_pin)), .count = 1 },
+            .in_base = @truncate(@intFromEnum(io_pin)),
+        },
+        .shift = .{
+            .out_shiftdir = .left,
+            .in_shiftdir = .left,
+            .autopull = true,
+            .autopush = true,
+        },
+    });
+
+    pio.sm_set_pindir(sm, @truncate(@intFromEnum(clk_pin)), 1, .out);
+    pio.sm_set_pindir(sm, @truncate(@intFromEnum(io_pin)), 1, .out);
+
+    pio.sm_set_pin(sm, @truncate(@intFromEnum(clk_pin)), 1, 0);
+    pio.sm_set_pin(sm, @truncate(@intFromEnum(io_pin)), 1, 0);
+
+    return .{ .pio = pio, .sm = sm, .cs_pin = cs_pin, .io_pin = io_pin, .clk_pin = clk_pin };
+}
+
+fn get_pin_pio_function(pio: hal.pio.Pio) hal.gpio.Function {
```

Comment:
this might be useful in the pio hal, or maybe the gpio hal?

Reply:
good idea, I will move it to pio hal